rand/rand.go

```diff
@@ -0,0 +1,194 @@
+	upperLettersLength           = len(upperLetters)
+	numbersAndLowerLettersLength = len(numberAndLowerLetters)
+	numbersAndUpperLettersLength = len(numbersAndUpperLetters)
+	numbersAndLettersLength      = len(numbersAndLetters)
+
+	charIdxBits = 6 // 6 bits to represent a letter index, for the biggest case, numbers and letters is 62
+	charIdxMask = 1<<charIdxBits - 1
+	charIdxMax  = 63 / charIdxBits // # of letter indices fitting in 63 bits
+	// this concurrent is chosen based on the benchmark result, if too small or too large, the performance will be bad
+	// the test result on a 4~56 cores container, the best range is from about 30 - around 100, choose 64 for it's easier to calculate
+	concurrent     = 64
+	concurrentMask = concurrent - 1
+)
+
+var (
+	mutexes        = make([]sync.Mutex, concurrent)
+	rands          = make([]rand.Source, concurrent)
+	index   uint32 = 0
+
+	optionMapping = map[int]string{
+		Numbers:                   numbers,
+		Letters:                   letters,
+		Letters | Lower | Upper:   letters,
+		Upper:                     upperLetters,
+		Lower:                     lowerLetters,
+		Numbers | Letters:         numbersAndLetters,
+		Numbers | Letters | Upper: numbersAndUpperLetters,
+		Numbers | Upper:           numbersAndUpperLetters,
+		Numbers | Letters | Lower: numberAndLowerLetters,
+		Numbers | Lower:           numberAndLowerLetters,
+	}
+
+	optionMappingLen = map[int]int{
+		Numbers:                   numbersLength,
+		Letters:                   lettersLength,
+		Letters | Lower | Upper:   lettersLength,
+		Upper:                     upperLettersLength,
+		Lower:                     lowerLettersLength,
+		Numbers | Letters:         numbersAndLettersLength,
+		Numbers | Letters | Upper: numbersAndUpperLettersLength,
+		Numbers | Upper:           numbersAndUpperLettersLength,
+		Numbers | Letters | Lower: numbersAndLowerLettersLength,
+		Numbers | Lower:           numbersAndLowerLettersLength,
+	}
+
+	optionNamesMapping = map[int]string{
+		Numbers:                   "Numbers",
+		Letters:                   "Letters",
+		Letters | Lower | Upper:   "Letters",
+		Upper:                     "Upper",
+		Lower:                     "Lower",
+		Numbers | Letters:         "NumbersAndLetters",
+		Numbers | Letters | Upper: "NumbersAndUpperLetters",
+		Numbers | Upper:           "NumbersAndUpperLetters",
+		Numbers | Letters | Lower: "NumberAndLowerLetters",
+		Numbers | Lower:           "NumberAndLowerLetters",
+	}
+)
+
+func init() {
+	for i := 0; i < concurrent; i++ {
+		rands[i] = rand.NewSource(time.Now().UnixNano() + rand.Int63())
+	}
+}
+
+// Int63 returns a non-negative pseudo-random 63-bit integer as an int64.
+func Int63() int64 {
+	current := atomic.AddUint32(&index, 1) & concurrentMask
+	mutexes[current].Lock()
+	defer mutexes[current].Unlock()
+	return rands[current].Int63()
+}
+
+// Uint64 returns a random uint64 value.
+func Uint64() uint64 {
+	r := Int63()
+	return uint64(r)>>31 | uint64(r)<<32
```

Comment:
I tried this and am getting an overflow [here](https://play.golang.org/p/Kia04aNLaFE).
Is there smth wrong?

Reply:
@ricenzca  It's a build error,  seems similar to https://stackoverflow.com/questions/29787422/whats-wrong-with-golang-constant-overflows-uint64

change `const` to `var` will work
https://play.golang.org/p/8pPVnN9H6UG